### PR TITLE
Allow injection of buffer overflow handler

### PR DESCRIPTION
**Pull request by edjames**
#### Summary
When the fluent-logger cannot connect to the fluent agent process, the contents of the internal buffer (`@pending`) are discarded when the buffer exceeds the buffer size limit. 

This change allows the optional injection of a handler which is invoked before the buffer is cleared. This allows that data to be handled rather than just thrown away.

#### Intended effect
If no `buffer_overflow_handler ` is provided in the config, there is no change in behaviour.

If a `buffer_overflow_handler` is provided it is invoked before clearing the contents of the internal buffer (`@pending`). The handler must implement a `flush` which accepts a single argument.

#### How I tested
I've added some specs for this, but I am also already running my fork in my production environment and it's working well.

#### Comments
Related issue: https://github.com/fluent/fluent-logger-ruby/issues/32

### CODE REVIEW

File: lib/fluent/logger/fluent_logger.rb [hunk 1]
```diff
@@ -107,6 +109,7 @@ def close
             rescue => e
               set_last_error(e)
               @logger.error("FluentLogger: Can't send logs to #{@host}:#{@port}: #{$!}")
+              flush_buffer_overflow_handler(@pending)
```
Comment (repeatedly) [inline lib/fluent/logger/fluent_logger.rb]: Should we check `@pending` size here?

Comment (repeatedly) [inline lib/fluent/logger/fluent_logger.rb]: Ah wait. This is in `close` method so no problem.

**Comment by repeatedly** [comment]
For easy to use, support `Proc` object is better.

```rb
Fluent::Logger::FluentLogger.new(nil, 
  :host => 'localhost', :port => 24224,
  :buffer_overflow_handler => Proc { |pendings| foo(pendings) })
```

In logger, we wrap this object.

**Comment by edjames** [comment]
Good point re Proc. 
Should I update this pull request to use a Proc instead of an object?

**Comment by repeatedly** [comment]
Supporting object is better for complex recover handling, e.g. use multiple variables.
So supporting both Proc and Object is better.
If passed object is a Proc object, we can wrap passed object using BufferOverflowHandlerWrapper object.

```rb
irb(main):003:0> (lambda {}).is_a?(Proc)
=> true
irb(main):004:0> (Proc.new {}).is_a?(Proc)
=> true
irb(main):005:0> class Wrapper; end
irb(main):006:0> (Wrapper.new).is_a?(Proc)
=> false
```

**Comment by edjames** [comment]
So are you suggesting support for both a Proc and an object?

In my opinion I think one or the other is cleaner, and I prefer your idea of a Proc (I should've thought of that!). I think if you support a Proc then you have all the flexibility you need and there is no real need to explicitly support an object too. I don't really like doing type-checking.

So I would say change this pull request to support a Proc. What do you think?

**Comment by repeatedly** [comment]
If you are OK with only Proc support, I'm also OK.
After update the patch, i will re-check and merge it.

**Comment by edjames** [comment]
Great, thanks a lot.

**Comment by repeatedly** [comment]
@edjames ping? Do you have a time to update the patch?

**Comment by edjames** [comment]
Hi, yeah, I will be doing the update in the next few days and then testing the change in my own production environment. The original patch is already in my production environment and I wouldn't want to make this change without knowing for sure that it works.

I'll be back soon for sure. Thanks!

**Comment by edjames** [comment]
I've updated the patch as discussed, changing to support for a Proc. I like this better than before. 
I'm using this new version in my own app and the behaviour is as expected.

**Comment by repeatedly** [comment]
Thanks!

**Pull request merged**